File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string serialize(TreeNode* node, unordered_map<string, int>& subtree,vector<TreeNode*>& duplicates){
        if (!node) return "#";
        string left=serialize(node->left,subtree,duplicates);
        string right=serialize(node->right,subtree,duplicates);
        string s=left+","+right+","+to_string(node->val);
        if(subtree[s]==1) duplicates.push_back(node);
        subtree[s]++;
        return s;
    }
    
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        unordered_map <string,int> subtree;
        vector<TreeNode*>duplicates;
        serialize(root,subtree,duplicates);
        return duplicates;
    }
};

File: 0652-find-duplicate-subtrees/0652-find-duplicate-subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def serialize(self, node, subtree, duplicates):
        if not node: return "null"
        l=self.serialize(node.left,subtree,duplicates)
        r=self.serialize(node.right,subtree,duplicates)
        s=l+","+r+","+str(node.val)
        if s in subtree:  
            if subtree[s]==1: duplicates.append(node)
            subtree[s]+=1
        else:
            subtree[s]=1
        return s
        
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        subtree={}
        duplicates=[]
        self.serialize(root,subtree,duplicates)
        return duplicates